akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala

```diff
@@ -325,26 +327,30 @@ object HttpRequest {
    * include a valid [[akka.http.scaladsl.model.headers.Host]] header or if URI authority and [[akka.http.scaladsl.model.headers.Host]] header don't match.
    */
   def effectiveUri(uri: Uri, headers: immutable.Seq[HttpHeader], securedConnection: Boolean, defaultHostHeader: Host): Uri = {
-    def findHost(headers: immutable.Seq[HttpHeader]): OptionVal[Host] = {
-      val it = headers.iterator
-      while (it.hasNext) it.next() match {
-        case h: Host ⇒ return OptionVal.Some(h)
-        case _       ⇒ // continue ...
-      }
-      OptionVal.None
-    }
-    val hostHeader: OptionVal[Host] = findHost(headers)
+    @tailrec def findHostAndWsUpgrade(it: Iterator[HttpHeader], host: OptionVal[Host] = OptionVal.None, wsUpgrade: Option[Boolean] = None): (OptionVal[Host], Boolean) =
```

Comment:
This way it returns earlier if `Upgrade.hasWebsocket == false` but that's rather uncommon. WDYT?